src/runtime/client-event-manager.js

```diff
@@ -91,37 +98,24 @@ export class ClientEventManager {
     this.filterers_.push(filterer);
   }
 
-
   /**
    * @overrides
    */
   logEvent(event) {
     validateEvent(event);
     this.lastAction_ = new Promise(resolve => {
-      for (let filterer = 0; filterer < this.filterers_.length; filterer++) {
-        if (this.filterers_[filterer](event) === FilterResult.CANCEL_EVENT) {
-          resolve();
-          return;
+      this.isReadyPromise_.then(() => {
```

Comment:
can you not do this.lastAction_ = this.isReadyPromise_.then(resolve => {

Reply:
The isReadyPromise is set to the configuredruntime promise which means the resolve function won't get passed (I think?); I believe it will be a ConfiguredRuntime object.  It does seem to work if I return Promise.resolve() instead of calling a passed resolve function from a new promise though.  I'm not very comfortable with promises yet and I'm not as sure it will work consistently so you will need to let me know about that.

Reply:
I did some testing of this and "resolve" becomes undefined for a large number of circumstances during testing.  I think you can't rely on the resolve function being passed in for some promise circumstances (such as during testing when something just passes in an already resolved promise and for promises where the resolution of the promise calls another promise with a specific value such as ConfiguredRuntime which I'm fairly certain is the case here).

So I think what I'm doing is safer and will work in all circumstances instead of an unknown subset.

Reply:
sure, here's one approach. why don't you tell me your thoughts on this and if required, we can modify further. Basically we are chaining promises here... 
this.lastAction_ = this.isReadyPromise_.then(() => {
                                  return Promise(resolve => {
                                      // filter events and call resolve()
                                      // OR for events not filtered, call listeneres and then call resolve()
                                   });
What this does is, when isReadyPromsie gets resolved, it creates another promise and assigns it to lastAction_. This should have no difference to your tests.

Another approach.
/**
 * @param {Event} event
 * @return {!Promise<boolean>}
 */
filterEvent(event) {
  let cancelEvent = false;
  for (let filterer = 0; filterer < this.filterers_.length; filterer++) {
        if (this.filterers_[filterer](event) === FilterResult.CANCEL_EVENT) {
          cancelEvent = true;
          break;
       }
  }
  return Promise.resolve(cancelEvent);
}

/**
   * @overrides
   */
  logEvent(event) {
    validateEvent(event);
    this.lastAction_ = this.isReadyPromise_.then(() => {
        return this.filterEvent(event);
      }).then(cancelled => {
        if (!cancelled) {
          // invoke listeners
        });
        return Promise.resolve();
      });
    });
I liked this second version better because this could even more asynchronous. Take a look and see how it may or may not impact your tests at all.

The bottom line is to ensure the code reflects our decision that everything waits on isReadyPromise to get resolved and hence lives in then of that.

Reply:
I've adjusted the code to eliminate the extra sub-promise and am returning Promise.resolve() instead.  I lean against having the chain of promises because I suspect that every bit of async probably adds overhead and that we should only do it if necessary or logical.  It feels superfluous to break it up further than it is.